<img src="http://oci02.img.iteso.mx/identidad_de_instancia_2018/ITESO/Logos%20ITESO/Logo-ITESO-Principal.jpg">

<center> <font color= #004A94> <font size = 6> Lic. Ingeniería Financiera </font> <br> <br> <font color= #047CFB> <font size = 4>I.F. Anehi Karely Mayorquin Higuera - if707468@iteso.mx </font>

<center><font color=#42c29b><font size=5> Laboratorio 2: Behavioral Finance </font> <br> <br>

<center> <font color= #047CFB> <font size = 4> Primavera 2020 </font>  |  <font color= #047CFB> <font size = 4> Repositorio: <a href='PONER LINK'>Link</https://github.com/a></font>

---

# Resumen:

En este laboratorio se realizara un análisis financiero a través de un algorítmo de trading. Dando datos historicos se generarán estadísticas básicas, medidas de atribución al desempeño en esta parte se análiza que tan bueno es el algorítmo. Por úlmo se análizaran los sesgos cognitivos que son errores sistemáticos que una persona comete en un proceso de pensamiento o de toma de decisión. 

# Codigo:

In [4]:
#Importacion de funciones y visualizaciones 
import funciones as fn
#import visualizaciones as vs  #Para las gráficas 
import plotly.graph_objects as go
import plotly.io as pio                           # renderizador para visualizar imagenes
pio.renderers.default = "browser"                 # render de imagenes para correr en script
import plotly.express as px
import pandas as pd
import principal as prin

# Parte 2) Realizar mediciones de estadísticas básicas de una cuenta de trading.

In [5]:
# Leer archivo 
data = fn.f_leer_archivo('archivo_tradeview_1.xlsx')

En esta parte podemos ver el archivo que se leyó anteriormente con sus datos históricos con el que se desarrollará el algoritmo de trading. Fueron descargados manualmente de Oanda y extraídos de Excel.  

In [6]:
data.head() 

order             opentime type  size  symbol   openprice       s/l  \
0  24316034  2019.08.27 06:57:50  buy  0.04  xauusd  1529.21000  1523.000   
1  24334215  2019.08.27 17:06:20  buy  0.01  eurjpy   117.66400   116.950   
2  24342155  2019.08.27 19:25:39  buy  0.05  eurjpy   117.34600   116.950   
3  24373260  2019.08.28 17:11:33  buy  0.01  eurusd     1.10791     1.104   
4  24392415  2019.08.29 06:35:27  buy  0.01  gbpusd     1.22024     1.208   

         t/p            closetime  closeprice  commission  taxes  swap  profit  
0  1543.0000  2019.08.27 09:16:01  1533.69000         0.0      0     0   17.92  
1   118.7000  2019.08.29 06:38:16   117.37400         0.0      0     0   -2.73  
2   119.7000  2019.08.28 21:01:24   117.45200         0.0      0     0    5.00  
3     1.1175  2019.08.29 06:38:28     1.10818         0.0      0     0    0.27  
4     1.2370  2019.08.29 06:37:23     1.21990         0.0      0     0   -0.34

In [7]:
# Columna de los tiempos
colum = fn.f_columna_tiempos(data)

In [8]:
# Columna de pips
colum_pips = fn.f_columna_pips(data)

In [9]:
#Estadisticas basicas y ranking
df_estadistic = fn.f_estadistica_ba(data)

A continuación, se mostrarán la serie de estadísticas básicas que se le aplicaron a los datos históricos. Podemos rescatar que en total solamente se ejecutaron 84 operaciones de las cuales 46 fueron ganadoras lo que viene siendo un poco más de la mitad 54.76% del total, del total de las ganadoras 20 fueron en compra y 26 en venta. Podemos ver que la media del profit es de 1.205 que, aunque no es muy alta sigue siendo positiva lo cual pues es bueno. Se observa también en los resultados de la tabla anterior, la ratio de efectividad es mayor al 50% lo cual significa como lo mencionamos anteriormente que más de la mitad de las operaciones fueron ganadoras.
 


In [10]:
df_estadistic['estadisticas']

Valor                             Descripcion
Ops totales            84                     Operaciones totales
Ganadoras              46                   Operaciones ganadoras
Ganadoras_c            20         Operaciones ganadoras de compra
Ganadoras_v            26          Operaciones ganadoras de venta
Perdedoras             38                  Operaciones perdedoras
Perdedoras_c           19        Operaciones perdedoras de compra
Perdedoras_v           19         Operaciones perdedoras de venta
Media (Profit)      1.205    Mediana de profit de las operaciones
Media (Pips)          3.6      Mediana de pips de las operaciones
r_efectividad    0.547619   Ganadoras Totales/Operaciones Totales
r_proporcion      1.21053   Ganadoras Totales/ Perdedoras Totales
r_efectividad_c  0.238095  Ganadoras Compras/ Operaciones Totales
r_efectividad_v  0.309524   Ganadoras Ventas/ Operaciones Totales

El ranking de divisas pone de ascendente a descendente las veces que dicha divisa hizo algún movimiento y cerro con ganancia tomando en cuenta la ratio de efectividad. Entonces lo que nos quiere decir la tabla siguiente es el porcentaje del total de movimientos las veces que ganaron cada divisa. Más adelante se mostrará un gráfico donde quedará más claro. 

In [11]:
df_estadistic['ranking']

ranking
btcusd  1.000000
eurgbp  1.000000
usdmxn  0.800000
xauusd  0.736842
gbpusd  0.666667
usdjpy  0.600000
eurusd  0.478261
audusd  0.333333
eurjpy  0.285714
usdcad  0.285714
gbpjpy  0.250000

In [12]:
data.head()

order            opentime type  size  symbol   openprice       s/l  \
0  24316034 2019-08-27 06:57:50  buy  0.04  xauusd  1529.21000  1523.000   
1  24334215 2019-08-27 17:06:20  buy  0.01  eurjpy   117.66400   116.950   
2  24342155 2019-08-27 19:25:39  buy  0.05  eurjpy   117.34600   116.950   
3  24373260 2019-08-28 17:11:33  buy  0.01  eurusd     1.10791     1.104   
4  24392415 2019-08-29 06:35:27  buy  0.01  gbpusd     1.22024     1.208   

         t/p           closetime  closeprice  commission  taxes  swap  profit  \
0  1543.0000 2019-08-27 09:16:01  1533.69000         0.0      0     0   17.92   
1   118.7000 2019-08-29 06:38:16   117.37400         0.0      0     0   -2.73   
2   119.7000 2019-08-28 21:01:24   117.45200         0.0      0     0    5.00   
3     1.1175 2019-08-29 06:38:28     1.10818         0.0      0     0    0.27   
4     1.2370 2019-08-29 06:37:23     1.21990         0.0      0     0   -0.34   

       time  pips  pips_acm  profit_acm  
0    8291.0  44.8      44.8       17.92  
1  135116.0 -29.0      15.8       15.19  
2   92145.0  10.6      26.4       20.19  
3   48415.0   2.7      29.1       20.46  
4     116.0  -3.4      25.7       20.12

# Parte 3) Calcular Medidas de Atribución al Desempeño expresadas semanalmente de una cuenta de trading y de la actividad del trader como persona.

Lo primero que se hará en esta parte será crear una función para estar y sumando/restando las ganancias/perdidas y realizar un acumulado de esto, tomando en cuenta que se comenzó con 5000, se agregó la última columna llamada "capital_acm" donde podemos ver como va aumentando el capital cuando se gana o se resta cuando se pierde.

In [14]:
#parte3
#Capital acumulado
cap_acum= fn.f_capital_acm(data)

In [15]:
data.head()

order            opentime type  size  symbol   openprice       s/l  \
0  24316034 2019-08-27 06:57:50  buy  0.04  xauusd  1529.21000  1523.000   
1  24334215 2019-08-27 17:06:20  buy  0.01  eurjpy   117.66400   116.950   
2  24342155 2019-08-27 19:25:39  buy  0.05  eurjpy   117.34600   116.950   
3  24373260 2019-08-28 17:11:33  buy  0.01  eurusd     1.10791     1.104   
4  24392415 2019-08-29 06:35:27  buy  0.01  gbpusd     1.22024     1.208   

         t/p           closetime  closeprice  commission  taxes  swap  profit  \
0  1543.0000 2019-08-27 09:16:01  1533.69000         0.0      0     0   17.92   
1   118.7000 2019-08-29 06:38:16   117.37400         0.0      0     0   -2.73   
2   119.7000 2019-08-28 21:01:24   117.45200         0.0      0     0    5.00   
3     1.1175 2019-08-29 06:38:28     1.10818         0.0      0     0    0.27   
4     1.2370 2019-08-29 06:37:23     1.21990         0.0      0     0   -0.34   

       time  pips  pips_acm  profit_acm  capital_acm  
0    8291.0  44.8      44.8       17.92      5017.92  
1  135116.0 -29.0      15.8       15.19      5015.19  
2   92145.0  10.6      26.4       20.19      5020.19  
3   48415.0   2.7      29.1       20.46      5020.46  
4     116.0  -3.4      25.7       20.12      5020.12

In [23]:
# Profits
df_profit = fn.f_profit_diario(data)

###### ESTADISTICAS METRICAS DE DESEMPEÑO

El índice sharpe Mide el rendimiento promedio excedente sobre la tasa libre de riesgo por unidad de riesgo incurrido.
En este caso el sharpe es de 0.16 lo que quiere decir que no es tan alto lo cual es bueno, pues en palabras sencillas es que se hizo mas rendimiento que el riesgo que hay.

El sortino_c y el sortino_v son bastantes similares al índice de Sharpe sin embargo, se toman en cuenta los datos de S&P500 ya que se "asume" que tiene un comportamiento parecido al de la risk free. Lo mismo sucede con Information Ratio. 

También tenemos el drawdown y drawup, que nos dicen que tan mal le va al portafolio cuando le va mal y que tan bien le va cuando le va bien. Toma en cuenta los puntos más bajos y forma una tendencia de estos, de la misma manera con los puntos más altos.



In [20]:
# Estadisticas de metricas de desempeño
df_profit_des = fn.f_estadisticas_mad(data)

In [21]:
df_profit_des['values']

Sharpe                                    0.166968
Sortino_c                                -0.136855
Sortino_v                                    0.153
Drawdown_capi    -190.48 | 2019-08-28 | 2019-09-03
Drawup_capi       421.83 | 2019-09-03 | 2019-09-25
Information                              0.0403861
Name: values, dtype: object

# Parte 4) Proponer, diseñar y calcular una función para obtener evidencia sobre la presencia de sesgos cognitivos en un trader.

En esta parte se realizarán los sesgos cognitivos, nos daremos cuenta en donde podremos estar cometiendo un error al estar cerrando alguna posición antes, o de manera contraria NO cerrando dicha posición teniendo el pensamiento erroneo de que está va a cambiar a nuestro favor.

In [23]:
#Parte4
# ocurrencias
sesgos = fn.f_be_de(data)

In [24]:
sesgos['resultados']

Valor
ocurrencias                     27
status_quo                0.592593
aversion_perdida           0.37037
sensibilidad_decreciente        no

Definiciones

- status_quo:  % de ocurrencias donde capital_perdedora/capital_acm < capital_ganadora/capital_acm.

- aversion_perdida: % de ocurrencias donde capital_perdedora/capital_ganadora es > 1.5 

- sensibilidad_decreciente: Comparar lo siguiente para la primera operación "ganadora" y la última operación "ganadora":
    Si el capital_acm de la cuenta aumentó.
    Si el capital_ganadora y/o capital_perdedora aumentaron (cualquiera de los dos, pero, al menos uno).
    Si el ratio capital_perdedora/capital_ganadora sigue siendo > 1.5.
    Si y sólo si todas las anteriores se cumplen, entonces, SI hay sensibilidad decreciente.

In [25]:
sesgos['ocurrencias']

[{'ocurrencia 1': {'timestamp': Timestamp('2019-08-27 09:16:01'),
   'operaciones': {'ganadora': {'instrumento': 'xauusd',
     'sentido': 'buy',
     'volumen': 0.04,
     'capital_ganadora': 17.92,
     'capital_acm': 5017.92},
    'perdedora': {'instrumento': 'eurjpy',
     'sentido': 'sell',
     'volumen': 0.05,
     'profit': 1.32,
     'capital_perdedora': -8.439}},
   'ratio_cp_capital_acm': 0.168,
   'ratio_cg_capital_acm': 0.357,
   'ratio_cp_cg': 0.471}},
 {'ocurrencia 2': {'timestamp': Timestamp('2019-08-28 21:01:24'),
   'operaciones': {'ganadora': {'instrumento': 'eurjpy',
     'sentido': 'buy',
     'volumen': 0.05,
     'capital_ganadora': 5.0,
     'capital_acm': 5020.19},
    'perdedora': {'instrumento': 'eurjpy',
     'sentido': 'buy',
     'volumen': 0.04,
     'profit': -9.98,
     'capital_perdedora': -6.402}},
   'ratio_cp_capital_acm': 0.128,
   'ratio_cg_capital_acm': 0.1,
   'ratio_cp_cg': 1.28}},
 {'ocurrencia 3': {'timestamp': Timestamp('2019-08-29 15:01:24'

# Gráficas

In [47]:
#%%
'''Gráfica 1: Ranking
+ Una gráfica de pastel (pie chart) con la información del dataframe que obtienes al mandar llamar la función df_1_ranking.

+ Leyenda en la parte derecha de la gráfica.

+ El valor más grande de la gráfica está "extraido" o "flotando" fuera del pastel

Nota: Revisa la sección de "Pie chart" y dentro busca la parte donde dice "Pulling sectors out from the center" en la documentación oficial en línea de plotly'''   
#Grafico de pastel 

rank = pd.DataFrame(prin.df_estadistic['ranking'])
rank['index']=rank.index
fig1 = go.Figure()
labels = rank.index
values = rank.ranking
fig1 = go.Figure(data=[go.Pie(labels=labels, values=values, pull=[0.2, 0.2, 0, 0, 0, 0, 0, 0, 0, 0, 0])])
fig1.update_layout(title='RANKING')
fig1.show()

In [48]:
#%%
'''Gráfica 2: DrawDown y DrawUp
+ Una gráfica de linea de color negro que represente la evolución del capital acumulado diario

+ Una línea recta punteada de color verde que represente el DrawUp, que va desde la fecha inicial hasta la fecha final de ocurrencia del máximo DrawUp.

+ Una línea recta punteada de color rojo que represente el DrawDown, que va desde la fecha inicial hasta la fecha final de ocurrencia del máximo DrawDown.'''
fig2 = go.Figure()
fig2.add_trace(go.Scatter(x=df_profit.timestamp, y=df_profit.profit_acm_d, mode = 'lines',
                             name = 'profit diario', line=dict(color='black')))
fig2.show()

NameError: name 'down' is not defined

In [ ]:
fig2.add_trace(go.Scatter(x=df_profit['timestamp'][down[1:]], y= df_profit['profit_acm'][down[1:]],mode='lines', name = 'drawdown', line=dict(color='Red',dash='dash')))
fig2.add_trace(go.Scatter(x=df_profit['timestamp'][up[1:]], y=df_profit['profit_acm'][up[1:]], mode='lines', name = 'drawup',line=dict(color='Green',dash='dash')))
fig2.update_layout(
        title="DrawUp y DrawDown",
        xaxis_title="Tiempo",
        yaxis_title="Portfolio acumulado diario",
        )
fig2.show()

In [27]:
#%%
'''Gráfica 3: Disposition Effect
+ Una gráfica de barras verticales

+ Hay 3 barras, cada una representa la cantidad de veces que se observó cada principio del disposition effect

+ Cada barra tendrá de nombre en el eje de las X los siguientes: status_quo, aversion_perdida, sensibilidad_decreciente'''
res = pd.DataFrame(sesgos['resultados'])
res['index1']=res.index
#%%
fig3 = go.Figure()
fig3 = px.bar(x=res.index1, y=res.Valor)
fig3.update_layout(
    title="Sesgos",
    xaxis_title="Variables",
    yaxis_title="cantidad/porcentage",
    )
fig3.show()

En la gráfica anterior se puenden observar 3 barras, la primera cuenta con el numero de ocurrencias que vienen siendo 27, la segunda es el status_quo el cual el 59% de las ocasiones la proporcion del capital perdedor entre el acumulado fue menor a la proporcion del capital ganador entre el acumulado, la tercera es la aversion_perdida que es el 37% lo que quiere decir que si es mayor a 1.5 lo cual se va a tomar en cuenta.